woocommerce/woocommerce-admin
# Throw custom exception in NoteTraits if notes are disabled

Fixes #6703.

When WC Admin is disabled, notes cannot be created using `NoteTraits` due to the `admin-note` data store not being registered.

This PR introduces a custom exception that is thrown in `NoteTraits` whenever WC Admin is disabled and any of the trait's methods that depend on the data store is called (as of this moment, these are all of the methods except for `wc_admin_active_for()`).

I did not add any test instructions to `/TESTING-INSTRUCTIONS.md` because it seems like this is not a kind of test that should be added there. If this is incorrect, please let me know!

### Detailed test instructions:

- Try to use any of the trait's methods during the `plugins_loaded` hook using `NoteTraits`.
  - Run the unit tests using `npm run test:php`, or
  - Place this code in `mu-plugins`:

  ```php
  add_action( 'plugins_loaded', function () {
  	// Disable WC Admin. If the line below is uncommented, no exception is thrown.
  	add_filter( 'woocommerce_admin_disabled','__return_true' );
  
  	// Try calling a NoteTraits method.
  	( new TestNoteTraitsUser() )->add_note();
  } );
  
  class TestNoteTraitsUser {
  	use Automattic\WooCommerce\Admin\Notes\NoteTraits;
  
  	// Constant required by NoteTraits.
  	const NOTE_NAME = 'test';
  
  	// Method required by NoteTraits.
  	public static function get_note() {
  		return new Automattic\WooCommerce\Admin\Notes\Note();
  	}
  
  	public function add_note() {
  		self::possibly_add_note();
  	}
  }
  ```
- See that a custom exception is thrown rather than a fatal error.



### Considered alternative approaches

Instead of checking if WC Admin is disabled in every `NoteTraits` method that depends on the `admin-note` data store being present, I also considered doing this:

```php
private static function load_data_store() {
	try {
		return new \WC_Data_Store::load( 'admin-note' );
	} catch ( Exception $e ) {
		throw new NotesDisabledException();
	}
}

// And then, in every method replace this:
$data_store = \WC_Data_Store::load( 'admin-note' );

// With this:
$data_store = self::load_data_store();
```

However, if notes become a feature that can be disabled/enabled (see #6568) or more reasons why note creation can fail appear (other than the data store not being registered), I think the approach suggested in this PR will be more flexible.
